frontend/packages/dev-console/src/components/buildconfig/sections/ImagesSection.tsx

```diff
@@ -0,0 +1,104 @@
+import { DropdownField, InputField } from '@console/shared';
+import { TriggersAndImageStreamFormData } from '../../edit-deployment/utils/edit-deployment-types';
+import ImageStream from '../../import/image-search/ImageStream';
+import FormSection from '../../import/section/FormSection';
+
+export type ImagesSectionFormData = {
+  formData: {
+    images: {
+      buildFrom: {
+        type: 'none' | 'imageStreamTag' | 'imageStreamImage' | 'dockerImage';
+        imageStreamTag: TriggersAndImageStreamFormData;
+        imageStreamImage: string;
+        dockerImage: string;
+      };
+      pushTo: {
+        type: 'none' | 'imageStreamTag' | 'imageStreamImage' | 'dockerImage';
+        imageStreamTag: TriggersAndImageStreamFormData;
+        imageStreamImage: string;
+        dockerImage: string;
+      };
+    };
+  };
+};
+
+const ImageOption: React.FC<{ fieldPrefix: string; label: string; dataTest: string }> = ({
+  fieldPrefix,
+  label,
+  dataTest,
+}) => {
+  const { t } = useTranslation();
+  const [{ value: type }] = useField(`${fieldPrefix}.type`);
+
+  const typeItems: Record<string, string> = {
+    none: t('devconsole~None'),
+    imageStreamTag: t('devconsole~Image Stream Tag'),
+    imageStreamImage: t('devconsole~Image Stream Image'),
+    dockerImage: t('devconsole~Docker image'),
+  };
+
+  return (
+    <div data-test={dataTest}>
+      <DropdownField
+        name={`${fieldPrefix}.type`}
+        label={label}
+        items={typeItems}
+        dataTest={`${dataTest} type`}
+      />
+
+      {type === 'imageStreamTag' ? (
+        <ImageStream
+          label={t('devconsole~Build from')}
+          formContextField={`${fieldPrefix}.imageStreamTag`}
+          dataTest={`${dataTest} image-stream-tag`}
+          required
+        />
+      ) : null}
+
+      {type === 'imageStreamImage' ? (
+        <InputField
+          label={t('devconsole~Image stream image')}
+          name={`${fieldPrefix}.imageStreamImage`}
+          type={TextInputTypes.text}
+          dataTest={`${dataTest} image-stream-image`}
+          required
+        />
+      ) : null}
+
+      {type === 'dockerImage' ? (
+        <InputField
+          label={t('devconsole~Docker image repository')}
+          name={`${fieldPrefix}.dockerImage`}
+          type={TextInputTypes.text}
+          dataTest={`${dataTest} docker-image`}
+          required
+        />
+      ) : null}
+    </div>
+  );
+};
+
+const ImagesSection: React.FC<{}> = () => {
+  const { t } = useTranslation();
+
+  return (
+    <FormSection title={t('devconsole~Images')} dataTest="section images">
+      <ImageOption
+        fieldPrefix="formData.images.buildFrom"
+        label={t('devconsole~Build from')}
+        dataTest="build-from"
+      />
+      <ImageOption
+        fieldPrefix="formData.images.pushTo"
+        label={t('devconsole~Push to')}
+        dataTest="push-to"
+      />
+    </FormSection>
+  );
+};
+
+export default ImagesSection;
```

Comment:
Make separate files for each component.

Reply:
Is there really a benefit when extracting this into another file? `ImagesSection` and `ImageOption` are currently closely related and we have "supporting sub" components in other code areas. But I can do this if this will block someone from giving an lgtm.

Reply:
I think it's okay to co-locate components that are tightly coupled in some cases.